File: frontend/src/types/book.ts
import { Author } from "./author";
import { HasTitleAndDescription } from "./base";
import { Category } from "./category";
import { Copy } from "./copy";
import { Publisher } from "./publisher";

export interface Book extends HasTitleAndDescription {
    categories: Category[] | null,
    authors: Author[] | null,
    publishers: Publisher[] | null,
    copies: Copy[] | null,
    copiesAvailable:number | null
};
export type PostBook = Omit<Book, "id" | "categories" | "authors" | "publishers" | "copies" | "copiesAvailable">;
export type Assign = {
    id:number,
    addId:number
}

File: Backend/backend/Models/Book.cs
﻿namespace Backend.Models;

using System.ComponentModel.DataAnnotations;
using System.ComponentModel.DataAnnotations.Schema;

public class Book : BaseModel
{
    [MinLength(2)]
    [MaxLength(50)]
    public string Title { get; set; } = null!;
    [MaxLength(256)]
    public string? Description { get; set; }
    public ICollection<Category> Categories { get; set; } = null!;
    public ICollection<Author> Authors { get; set; } = null!;  
    public ICollection<Copy> Copies { get; set; } = null!;
    [NotMapped]
    public ICollection<Publisher>? Publishers { get => Copies?.Select(c => c.Publisher).Distinct().ToList(); }
    [NotMapped]
    public int CopiesAvailable { get => Copies == null ? 0 : Copies.Where(copy => copy.IsAvailable).Count(); }
}
